0805 2회차 김광표 모의고사

Pangpyo opened the pull request.

hphktutor3 commented on 2회차/김광표/0_민석이의과제체크하기.py:
```diff
@@ -1,3 +1,15 @@
+    stu_ass = map(int, input().split()) # 숙제를 제출한 학생을 입력받는다.
+    for j in stu_ass : # 모든 학생들의 리스트에서 숙제를 제출한 학생의 번호를 지워준다.
+        stu.remove(j)
+    print('#%d'%i, *stu)
```
접근법과 코드가 아주 깔끔하고 정확하네요
다양한 풀이법을 고민해보자면
for I in range(1,N+1) : 를 순회하면서
stu_ass에 i가 없으면 정답으로 출력하는 방법도 있겠네요.
여러 풀이법을 참고하시면 도움이 되실 수도 있어서 공유드립니다 잘하셨습니다 😄

hphktutor3 commented on 2회차/김광표/1_조교의성적매기기.py:
```diff
@@ -1,3 +1,26 @@
+
+    ans = grade[int((rank)//(N/10))] 
+    # 등급은 총 10가지이므로 (총 학생수/10)으로 등수를 나눈 몫을 평점 리스트에 대입하면 평점이 나온다.
+    print('#%d'%i, ans)
```
딕셔너리를 활용한 풀이와 grade를 리스트로 만들어서 바로바로 평점을 계산하는 풀이가 인상적입니다.
students를 딕셔너리가 아닌 리스트를 활용한 풀이로 하신다면 어떨지 고민해보셔도 좋을 것 같네요.
딕셔너리의 정렬법에서 key를 활용한 풀이가 너무 좋네요 응용력이 대단하신 것 같습니다.
다만 등수를 계산하는 과정에서 index()를 활용하셨다면 조금 더 코드가 간결해질 수 있을 것 같네요

hphktutor3 commented on 2회차/김광표/2_암호생성기.py:
```diff
@@ -1,3 +1,25 @@
+                break
+        if point == 1 :
+            break
+    print('#%d'%i, *password)
```
암호생성기는 어려울 수 있는 문제인데 너무 잘 해결하셨네요.
한가지 아쉬운 부분은 이미 알고 계실 수도 있겠지만 password는 왼쪽에서 데이터가 빠져나가서 오른쪽으로 데이터가 들어가고 있습니다.
해당 작업에 가장 적합한 자료구조는 queue / deque 가 되겠네요. 이들을 활용해서 코드를 작성한다면 어떻게 될지 고민해보시면 좋을 것 같습니다. 생각보다 pop(0) 을 하는 작업이 오래걸리는 함수이기 때문에 가능하다면 deque를 사용하시는 것을 추천드립니다.

hphktutor3 commented on 2회차/김광표/3_파리퇴치.py:
```diff
@@ -1,3 +1,23 @@
+        if max(row) > ans :
+            ans = max(row)
+
+    print('#%d'%i, ans)
```
catch를 따로 설정해서 값을 측정하는 부분이 인상적입니다.
더 나아가 범위를 너무 잘 설정하셔서 범위를 벗어나는지 굳이 확인하지 않아도 되는 코드 구현이
아주 깔끔하네요 👍 👍 
한가지 팁을 드리자면 map의 응용으로 ans = max(map(max, catch))) 이렇게 작성하시면
이차원리스트의 최대값을 구하실 수 있습니다. sum또는 max에서 자주 사용되는 유형이니
기억해두시면 은근 편할 것 같네요

hphktutor3 commented on 2회차/김광표/4_어디에단어가들어갈수있을까.py:
```diff
@@ -1,3 +1,31 @@
+                if col == K :
+                    ans += 1
+                col = 0
+    print('#%d'%i, ans)
```
가로와 세로를 하나의 for문에서 구현하시는 모습이 너무 좋습니다.
다른 분들께서는 코드가 헷갈릴 수 있어서 따로 쓰시기도 하시지만
이렇게 익숙하게 잘 활용하시는 것을 보니 이차원 리스트와 반복문에 충분히 자신감을 가지셔도 좋을 것 같습니다.
크게 코멘트 드릴 부분은 없네요. 너무 잘하셨습니다 👍

hphktutor3 commented on 2회차/김광표/5_괄호짝짓기.py:
```diff
@@ -1,3 +1,25 @@
+            else : # 아닐 경우 유효하지 않으므로 0을 ans에 넣어주고 for문 break
+                ans = 0
+                break
+    print('#%d'%i, ans)
```
스택을 활용한 괄호 문제 해결이 아주 좋습니다.
역시 수업시간에 비슷한 문제를 해서 그런지 아주 깔끔하게 작성해주셨네요 👍 
코드에서 index를 활용해 짝이 맞는지 확인하셨는데, 딕셔너리를 통해 짝이 맞는지 확인하기 위해서는 어떻게 해야할 지
한번 고민해보시고 직접 구현해보신다면 더욱 좋겠네요. 고생 많으셨습니다!!

Pangpyo commented:
정성스러운 피드백 감사합니다~!

2022년 8월 11일 (목) 오후 7:23, hphktutor3 ***@***.***>님이 작성:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In 2회차/김광표/0_민석이의과제체크하기.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#discussion_r943317609>:
>
> >  import sys
>
>  sys.stdin = open("_민석이의과제체크하기.txt")
> +
> +T = int(input())
> +
> +for i in range(1, T+1) :
> +    N, K = map(int, input().split())
> +    stu = list(range(1, N+1)) # 모든 학생들의 리스트를 번호로 만든다.
> +    stu_ass = map(int, input().split()) # 숙제를 제출한 학생을 입력받는다.
> +    for j in stu_ass : # 모든 학생들의 리스트에서 숙제를 제출한 학생의 번호를 지워준다.
> +        stu.remove(j)
> +    print('#%d'%i, *stu)
>
> 접근법과 코드가 아주 깔끔하고 정확하네요
> 다양한 풀이법을 고민해보자면
> for I in range(1,N+1) : 를 순회하면서
> stu_ass에 i가 없으면 정답으로 출력하는 방법도 있겠네요.
> 여러 풀이법을 참고하시면 도움이 되실 수도 있어서 공유드립니다 잘하셨습니다 😄
> ------------------------------
>
> In 2회차/김광표/1_조교의성적매기기.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#discussion_r943320113>:
>
> > +    students = {}
> +    for j in range(1, N+1) :
> +        mid, final, assin = list(map(int, input().split())) # 중간고사, 기말고사, 과제 점수를 입력받는다
> +        score = mid*0.35 + final*0.45 +assin*0.20 # 중간고사, 기말고사, 과제점수를 비중에 맞게 총 점수를 계산한다.
> +        students[j] = score # 학생번호 : 점수의 딕셔너리를 만든다.
> +    students = sorted(students.items(), key = lambda x : -x[1]) # 딕셔너리를 점수의 내림차순으로 정렬한다.
> +    rank = 0 # 등수를 저장할 변수를 만든다. ( 0등이 1등이다.)
> +    for k,v in students :
> +        if k == K : # 내가 찾는 학생이 몇 등에 있는지 확인한다.
> +            break
> +        rank += 1
> +    grade = ['A+', 'A0', 'A-', 'B+', 'B0', 'B-', 'C+', 'C0', 'C-', 'D0'] # 평점을 리스트로 만들었다.
> +
> +    ans = grade[int((rank)//(N/10))]
> +    # 등급은 총 10가지이므로 (총 학생수/10)으로 등수를 나눈 몫을 평점 리스트에 대입하면 평점이 나온다.
> +    print('#%d'%i, ans)
>
> 딕셔너리를 활용한 풀이와 grade를 리스트로 만들어서 바로바로 평점을 계산하는 풀이가 인상적입니다.
> students를 딕셔너리가 아닌 리스트를 활용한 풀이로 하신다면 어떨지 고민해보셔도 좋을 것 같네요.
> 딕셔너리의 정렬법에서 key를 활용한 풀이가 너무 좋네요 응용력이 대단하신 것 같습니다.
> 다만 등수를 계산하는 과정에서 index()를 활용하셨다면 조금 더 코드가 간결해질 수 있을 것 같네요
> ------------------------------
>
> In 2회차/김광표/2_암호생성기.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#discussion_r943323600>:
>
> > +    except :
> +        break
> +    password = list(map(int, input().split()))
> +    point = 0 # break 포인트 설정
> +    while 1 :
> +        for j in range(1, 6) : #1~5까지의 범위 설정
> +            a = password.pop(0) # 첫번째 인덱스를 pop해줌
> +            a -= j # pop 해준 값에 j를 빼줌(1~5)
> +            password.append(a) # 다시 그 값을 리스트에 append해줌
> +            if a <= 0 : # 그 과정에서 a가 0 이하가 되면 리스트의 마지막 값을 0으로 만든 뒤 브레이크해줌
> +                password[-1] = 0
> +                point = 1 # 바깥쪽의 while문도 break해주기 위한 변수
> +                break
> +        if point == 1 :
> +            break
> +    print('#%d'%i, *password)
>
> 암호생성기는 어려울 수 있는 문제인데 너무 잘 해결하셨네요.
> 한가지 아쉬운 부분은 이미 알고 계실 수도 있겠지만 password는 왼쪽에서 데이터가 빠져나가서 오른쪽으로 데이터가 들어가고
> 있습니다.
> 해당 작업에 가장 적합한 자료구조는 queue / deque 가 되겠네요. 이들을 활용해서 코드를 작성한다면 어떻게 될지 고민해보시면
> 좋을 것 같습니다. 생각보다 pop(0) 을 하는 작업이 오래걸리는 함수이기 때문에 가능하다면 deque를 사용하시는 것을 추천드립니다.
> ------------------------------
>
> In 2회차/김광표/3_파리퇴치.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#discussion_r943325439>:
>
> > +
> +for i in range(1, T+1) :
> +    N, M = map(int, input().split()) # 파리가 있는 영역의 넓이와 파리채의 넓이
> +    flys = [list(map(int, input().split())) for _ in range(N)] # 파리가 있는 영역의 각 파이들의 수
> +    catch = [[0]*(N-M+1) for _ in range(N-M+1)] # 각 위치별 파리채질 한번으로 잡을 수 있는 파리의 수를 저장할 행렬
> +    for x in range(N-M+1) :
> +        for y in range(N-M+1) : # N*N의 영역에서는 총 (N-M+1)*(N-M+1) 만큼의 위치에서 파리채질을 할 수 있다.
> +            for a in range(x, x+M) : # 각 파리채질마다 잡을 수 있는 파리의 수를 구한다.
> +                for b in range(y, y+M) :
> +                    catch[x][y] += flys[a][b]
> +    ans = 0
> +    for row in catch : # 가장 많은 파리를 잡을 수 있는 경우의 파리 수를 구한다.
> +        if max(row) > ans :
> +            ans = max(row)
> +
> +    print('#%d'%i, ans)
>
> catch를 따로 설정해서 값을 측정하는 부분이 인상적입니다.
> 더 나아가 범위를 너무 잘 설정하셔서 범위를 벗어나는지 굳이 확인하지 않아도 되는 코드 구현이
> 아주 깔끔하네요 👍 👍
> 한가지 팁을 드리자면 map의 응용으로 ans = max(map(max, catch))) 이렇게 작성하시면
> 이차원리스트의 최대값을 구하실 수 있습니다. sum또는 max에서 자주 사용되는 유형이니
> 기억해두시면 은근 편할 것 같네요
> ------------------------------
>
> In 2회차/김광표/4_어디에단어가들어갈수있을까.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#discussion_r943326555>:
>
> > +        for y in range(N) :
> +            if puzzl[x][y] == 1 : # 1일경우 빈 칸이므로 row에 +1
> +                row += 1
> +            if puzzl[x][y] == 0 or y == N-1:
> +            # 0일경우나 가장 끝 인덱스를 검사하는 경우 이전까지의 빈칸의 개수를 세어준 후,
> +            # 단어의 길이와 같으면 답에 +1
> +                if row == K :
> +                    ans += 1
> +                row = 0
> +            if puzzl[y][x] == 1 : # 열에 대해서도 같은 작업 반복
> +                col += 1
> +            if puzzl[y][x] == 0 or y == N-1:
> +                if col == K :
> +                    ans += 1
> +                col = 0
> +    print('#%d'%i, ans)
>
> 가로와 세로를 하나의 for문에서 구현하시는 모습이 너무 좋습니다.
> 다른 분들께서는 코드가 헷갈릴 수 있어서 따로 쓰시기도 하시지만
> 이렇게 익숙하게 잘 활용하시는 것을 보니 이차원 리스트와 반복문에 충분히 자신감을 가지셔도 좋을 것 같습니다.
> 크게 코멘트 드릴 부분은 없네요. 너무 잘하셨습니다 👍
> ------------------------------
>
> In 2회차/김광표/5_괄호짝짓기.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#discussion_r943327846>:
>
> > +    leftpars = ['{','[','<','('] # 좌괄호
> +    rightpars = ['}',']','>',')'] # 우괄호
> +    parcheck = [] # 괄호들이 들어가고 지워질 리스트
> +    ans = 1
> +    for par in pars :
> +        if par in leftpars : # 좌괄호가 나올 경우 parcheck에 append해줌
> +            parcheck.append(par)
> +        else : # 우괄호가 나올 경우
> +            if rightpars.index(par) == leftpars.index(parcheck[-1]) :
> +                # parcheck의 가짱 끝에 해당 우괄호과 짝이 되는 좌괄호가 있는지 체크
> +                parcheck.pop()
> +                # 맞을 경우 짝이 되는 괄호를 pop해줌
> +            else : # 아닐 경우 유효하지 않으므로 0을 ans에 넣어주고 for문 break
> +                ans = 0
> +                break
> +    print('#%d'%i, ans)
>
> 스택을 활용한 괄호 문제 해결이 아주 좋습니다.
> 역시 수업시간에 비슷한 문제를 해서 그런지 아주 깔끔하게 작성해주셨네요 👍
> 코드에서 index를 활용해 짝이 맞는지 확인하셨는데, 딕셔너리를 통해 짝이 맞는지 확인하기 위해서는 어떻게 해야할 지
> 한번 고민해보시고 직접 구현해보신다면 더욱 좋겠네요. 고생 많으셨습니다!!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kdt-hphk/01-PJT-04/pull/36#pullrequestreview-1069480707>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZ46ITCUL42F3RPIGTT5B2LVYTICPANCNFSM55VETYHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
